clients/java/src/main/java/io/camunda/zeebe/client/impl/worker/JobWorkerImpl.java

```diff
@@ -234,24 +239,89 @@ private void getPollInterval() {
   }
 
   private void handleJob(final ActivatedJob job) {
+    final boolean notEnqueued = !tryToEnqueueJob(job);
+    if (notEnqueued) {
+      // TODO logic to handle the faults?
+      // The job is likely to timed out anyway, so we potentially just skip it and don't need to do
+      // any further error handling
+
+      // Interesting is here whether we maybe want to explode, since adding other jobs from the
+      // batch also doesn't make much sense.
+      return;
+    }
+
     metrics.jobActivated(1);
-    executor.execute(jobHandlerFactory.create(job, this::handleJobFinished));
+    scheduleConsumeJob(this::handleJobFinished);
+  }
+
+  private boolean tryToEnqueueJob(final ActivatedJob job) {
+    boolean notEnqueued;
+    try {
+      notEnqueued = jobsQueue.offer(job, 10, TimeUnit.SECONDS);
+    } catch (final InterruptedException e) {
+      notEnqueued = false;
+    }
+    return notEnqueued;
   }
 
   private void handleStreamedJob(final ActivatedJob job) {
+    final boolean notEnqueued = !tryToEnqueueJob(job);
+    if (notEnqueued) {
+      // TODO logic to handle the faults?
+      // The job is likely to timed out anyway, so we potentially just skip it and don't need to do
+      // any further error handling
+      return;
+    }
+
     metrics.jobActivated(1);
-    executor.execute(jobHandlerFactory.create(job, this::handleStreamJobFinished));
+    scheduleConsumeJob(this::handleStreamJobFinished);
+  }
+
+  private void consumeJobFromQueue(final Runnable finalizer) {
+    if (jobsQueue.peek() == null) {
+      // stop if there is nothing - we will triggered by push/poll again
+      return;
+    }
+
+    ActivatedJob job;
+    try {
+      job = jobsQueue.poll(10, TimeUnit.SECONDS);
+      // TODO should we really do a poll ? Maybe a peek and remove it later after handling?
+    } catch (final InterruptedException ie) {
```

Comment:
❓ If we remove the job from the queue in order to work on it, this means we might receive one more. Could be ok or not. This would mean we would have a capacity of jobs plus count of threads handling currently a job as actual pending job size. I guess we need to make a decision what we want here